src/DeviceInput/deviceInputSystem.ts

```diff
@@ -291,13 +295,26 @@ export class DeviceInputSystem implements IDisposable {
             if (kbKey) {
                 kbKey[evt.keyCode] = 0;
                 if (this.onInputChanged) {
-                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode]);
+                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode], evt);
+                }
+            }
+        });
+
+        this._keyboardBlurEvent = ((evt) => {
+            if (this._keyboardActive) {
+                const kbKey = this._inputs[DeviceType.Keyboard][0];
+
+                for (let i = 0; i < kbKey.length; i++) {
+                    if (kbKey[i] !== 0) {
+                        kbKey[i] = 0;
+                    }
                 }
             }
         });
 
-        window.addEventListener("keydown", this._keyboardDownEvent);
-        window.addEventListener("keyup", this._keyboardUpEvent);
+        this._elementToAttachTo.addEventListener("keydown", this._keyboardDownEvent);

```

Comment:
without defining `tabindex` on an html element, the window would be the only element to attach to. `window` of `document.body`, actually. But if you try to attach a keyboard event to a canvas (without setting its `tabindex`), events won't go through. Is that accounted for here?
It will work in the playground, as we are adding tabindex to the canvas.

This relates to keyboard events only. blur should be on the canvas element itself.

Reply:
Talked offline, adding line to add engine's tabIndex to elementToAttachTo if current one is -1

Reply:
Resolved with setting `.tabIndex`